control_plane/attachment_service/src/service.rs

```diff
@@ -2115,24 +2116,66 @@ impl Service {
         for (tenant_shard_id, node) in targets {
             let client = mgmt_api::Client::new(node.base_url(), self.config.jwt_token.as_deref());
             futs.push(async move {
-                let result = client.tenant_secondary_download(tenant_shard_id).await;
-                (result, node)
+                let result = client
+                    .tenant_secondary_download(tenant_shard_id, wait)
+                    .await;
+                (result, node, tenant_shard_id)
             })
         }
 
         // Handle any errors returned by pageservers.  This includes cases like this request racing with
         // a scheduling operation, such that the tenant shard we're calling doesn't exist on that pageserver any more, as
         // well as more general cases like 503s, 500s, or timeouts.
-        while let Some((result, node)) = futs.next().await {
-            let Err(e) = result else { continue };
-
-            // Secondary downloads are always advisory: if something fails, we nevertheless report success, so that whoever
-            // is calling us will proceed with whatever migration they're doing, albeit with a slightly less warm cache
-            // than they had hoped for.
-            tracing::warn!("Ignoring tenant secondary download error from pageserver {node}: {e}",);
+        let mut aggregate_progress = SecondaryProgress::default();
+        let mut aggregate_status: Option<StatusCode> = None;
+        let mut error: Option<mgmt_api::Error> = None;
+        while let Some((result, node, tenant_shard_id)) = futs.next().await {
+            match result {
+                Err(e) => {
+                    // Secondary downloads are always advisory: if something fails, we nevertheless report success, so that whoever
+                    // is calling us will proceed with whatever migration they're doing, albeit with a slightly less warm cache
+                    // than they had hoped for.
```

Comment:
The comment in `secondary.rs` says that this request is mostly a "hook for tests", but you're alluding to a migration procedure. Could you clarify when this request would actually be called? Secondary downloads should just happen in the background anyways, right?

Reply:
Ah, right. The storage controller uses it for live migrations. Why though? I expected that in the case of a live migration we want to prioritise getting attached to the secondary tenant asap, but instead we wait up to 300 seconds for the warm-up. Whatever was already downloaded in the background should be good enough, no?

Reply:
>The comment in secondary.rs says that this request is mostly a "hook for tests",

Good catch, that's outdated.  Fixed 761254978.

> I expected that in the case of a live migration we want to prioritise getting attached to the secondary tenant asap

Sometimes, we might.  But for migrations we're doing outside an emergency (e.g. when we migrate shards around after a tenant split, or in preparation for a node upgrade), we should make every effort to give the user the least possible disruption.

In the storage controller, we always do this "nicest possible migration" process when the origin node is available, and a "hard migration" (i.e. just attach ASAP) if the origin node isn't available.  There should be more nuance there in future: e.g. we might do a migration where the origin is up but overloaded, in which case we should not be so generous with time waiting for downloads.